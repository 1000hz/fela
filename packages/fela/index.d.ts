/* TODO: Add all Interfaces and Methods */
interface Renderer {
    renderRule(rule: Function, props?: any): string
}

export declare function createRenderer(config?:any): Renderer;
export declare function combineRules(...rules:any[]): Function;
export declare function enhance(...enhancers:any[]): Function;
export declare function render(renderer:any, mountNode:HTMLElement): void;
