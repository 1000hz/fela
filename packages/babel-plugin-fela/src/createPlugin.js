import { generateMonolithicClassName, arrayReduce, arrayEach } from 'fela-utils'

const defaultConfig = {
  precompile: false
}

export default function createPlugin(config = {}) {
  const configWithDefaults = {
    ...defaultConfig,
    ...config
  }

  return ({ types: t, traverse }) => {
    // helper method to extract static style properties from AST objects
    function extractStaticStyle(props, path) {
      const staticStyle = arrayReduce(
        props,
        (style, node) => {
          if (
            !node.shorthand &&
            (t.isStringLiteral(node.value) || t.isNumericLiteral(node.value))
          ) {
            style.push(node)
          } else if (t.isObjectExpression(node.value)) {
            style.push(
              t.objectProperty(
                node.key,
                t.objectExpression(extractStaticStyle(node.value.properties, path))
              )
            )
          }

          return style
        },
        []
      )

      arrayEach(staticStyle, prop => {
        if (t.isObjectExpression(prop.value)) {
          if (prop.value.properties.length === 0) {
            props.splice(props.indexOf(prop), 1)
          }
        } else {
          props.splice(props.indexOf(prop), 1)
        }
      })

      return staticStyle
    }

    // helper to transform AST ObjectExpressions into JS objects
    function createStaticJSObject(props, path) {
      return props.reduce((obj, node) => {
        if (!node.shorthand && (t.isStringLiteral(node.value) || t.isNumericLiteral(node.value))) {
          obj[node.key.name] = node.value.value
        } else if (t.isObjectExpression(node.value)) {
          obj[node.key.value] = createStaticJSObject(node.value.properties, path)
        }

        return obj
      }, {})
    }

    function getRuleScope(path) {
      const rule = path.node.arguments[0]
      let ruleDeclaration,
        functionExpression

      if (t.isIdentifier(rule)) {
        if (path.scope.hasBinding(rule.name)) {
          const binded = path.scope.bindings[rule.name].path

          if (t.isVariableDeclarator(binded) || t.isFunctionDeclaration(binded)) {
            ruleDeclaration = binded
            functionExpression = t.isVariableDeclarator(binded)
              ? ruleDeclaration.node.init
              : ruleDeclaration.node
          }
        }
      } else if (t.isFunctionDeclaration(rule)) {
        ruleDeclaration = rule
        functionExpression = rule.node
      } else if (t.isArrowFunctionExpression(rule) || t.isFunctionExpression(rule)) {
        ruleDeclaration = rule
        functionExpression = rule
      }

      if (t.isArrowFunctionExpression(functionExpression)) {
        if (t.isObjectExpression(functionExpression.body)) {
          functionExpression.body = t.blockStatement([t.returnStatement(functionExpression.body)])

          if (t.isVariableDeclarator(ruleDeclaration)) {
            ruleDeclaration.node.init = functionExpression
          } else {
            ruleDeclaration.node = functionExpression
          }
        }
      }

      return {
        ruleDeclaration,
        functionExpression
      }
    }

    return {
      visitor: {
        CallExpression(path, parentPath) {
          if ((path.node.callee.name = 'createComponent')) {
            const { ruleDeclaration, functionExpression } = getRuleScope(path)

            if (!ruleDeclaration || !functionExpression) {
              return false
            }

            let didTraverse
            let renderer

            // pass renderer as second rule parameter if not already happening
            if (functionExpression.params && functionExpression.params.length === 0) {
              functionExpression.params.push('_')
            }

            if (functionExpression.params && functionExpression.params.length === 2) {
              if (!t.isIdentifier(functionExpression.params[1])) {
                return
              }

              if (functionExpression.params[1].name !== 'renderer') {
                renderer = functionExpression.params[1].name
              }
            }

            if (functionExpression.params && functionExpression.params.length === 1) {
              functionExpression.params.push(t.identifier('renderer'))
              renderer = 'renderer'
            }

            const traverser = {
              ObjectExpression(childPath) {
                if (!didTraverse) {
                  childPath.node.root = true
                  didTraverse = true
                }

                if (childPath.node.root) {
                  const props = childPath.node.properties

                  let id,
                    blockBody
                  let staticStyle = extractStaticStyle(props, childPath)

                  // static style precompilation
                  if (configWithDefaults.precompile && configWithDefaults.renderer) {
                    const jsObject = createStaticJSObject(staticStyle)

                    const className = configWithDefaults.renderer.renderRule(() => jsObject)

                    if (className) {
                      id = className.replace(/ /g, '')
                      blockBody = t.blockStatement([
                        t.expressionStatement(
                          t.assignmentExpression(
                            '=',
                            t.memberExpression(
                              t.memberExpression(t.identifier(renderer), t.identifier('cache')),
                              t.identifier(id)
                            ),
                            t.stringLiteral(className)
                          )
                        )
                      ])
                    }
                    staticStyle = {}
                  }

                  // simple static style prerendering
                  if (Object.keys(staticStyle).length > 0) {
                    id = generateMonolithicClassName(staticStyle)
                    blockBody = t.blockStatement([
                      t.expressionStatement(
                        t.assignmentExpression(
                          '=',
                          t.memberExpression(
                            t.memberExpression(t.identifier(renderer), t.identifier('cache')),
                            t.identifier(id)
                          ),
                          t.callExpression(
                            t.memberExpression(t.identifier(renderer), t.identifier('renderRule')),
                            [t.ArrowFunctionExpression([], t.objectExpression(staticStyle))]
                          )
                        )
                      )
                    ])
                  }

                  if (id && blockBody) {
                    childPath.node.properties.unshift(
                      t.objectProperty(
                        t.identifier('_className'),
                        t.memberExpression(
                          t.memberExpression(t.identifier(renderer), t.identifier('cache')),
                          t.identifier(id)
                        )
                      )
                    )

                    let isRoot

                    traverse(
                      functionExpression,
                      {
                        BlockStatement(bPath) {
                          if (!isRoot) {
                            isRoot = true

                            bPath.node.body.unshift(
                              t.ifStatement(
                                t.unaryExpression(
                                  '!',
                                  t.memberExpression(
                                    t.memberExpression(
                                      t.identifier(renderer),
                                      t.identifier('cache')
                                    ),
                                    t.identifier(id)
                                  )
                                ),
                                blockBody
                              )
                            )
                          }
                        }
                      },
                      childPath.scope,
                      childPath
                    )
                  }
                }
              }
            }

            if (ruleDeclaration.traverse) {
              ruleDeclaration.traverse(traverser)
            } else {
              traverse(ruleDeclaration, traverser, path.scope, path)
            }
          }
        }
      }
    }
  }
}
