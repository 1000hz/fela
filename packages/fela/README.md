# fela

NPM package to get access to Fela's API.<br>
Check out the [repository](https://github.com/rofrischmann/fela) for further information.
