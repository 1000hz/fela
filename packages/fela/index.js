module.exports = require('./lib/fela')
