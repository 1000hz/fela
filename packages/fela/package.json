{
  "name": "fela",
  "version": "5.2.0",
  "description": "Universal, Dynamic & High-Performance Styling in JavaScript",
  "typings": "index.d.ts",
  "main": "lib/index.js",
  "module": "es/index.js",
  "jsnext:main": "es/index.js",
  "files": [
    "LICENSE",
    "README.md",
    "index.d.ts",
    "lib/**",
    "es/**"
  ],
  "repository": "https://github.com/rofrischmann/fela/",
  "keywords": [
    "fela",
    "dynamic styling",
    "stylesheet",
    "css",
    "styling",
    "cssinjs"
  ],
  "author": "Robin Frischmann",
  "license": "MIT",
  "dependencies": {
    "css-in-js-utils": "^2.0.0",
    "fela-utils": "^6.0.1",
    "recompose": "^0.25.0"
  }
}
