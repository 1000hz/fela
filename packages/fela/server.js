module.exports = require('lib/felaServer')
