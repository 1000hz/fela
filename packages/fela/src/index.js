/* @flow */
import createRenderer from './createRenderer'
import combineRules from './combineRules'
import enhance from './enhance'

// bindings
import connectFactory from './bindings/connectFactory'
import createComponentFactory from './bindings/createComponentFactory'
import ProviderFactory from './bindings/ProviderFactory'
import ThemeProviderFactory from './bindings/ThemeProviderFactory'
import withThemeFactory from './bindings/withThemeFactory'
import createTheme from './bindings/createTheme'

export {
  createRenderer,
  combineRules,
  enhance,
  connectFactory,
  createComponentFactory,
  ProviderFactory,
  ThemeProviderFactory,
  withThemeFactory,
  createTheme
}
