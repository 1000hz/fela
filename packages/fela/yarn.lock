# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


css-in-js-utils@2.0.0:
  version "2.0.0"
  resolved "https://registry.yarnpkg.com/css-in-js-utils/-/css-in-js-utils-2.0.0.tgz#5af1dd70f4b06b331f48d22a3d86e0786c0b9435"
  dependencies:
    hyphenate-style-name "^1.0.2"

css-in-js-utils@^2.0.0:
  version "2.0.1"
  resolved "https://registry.yarnpkg.com/css-in-js-utils/-/css-in-js-utils-2.0.1.tgz#3b472b398787291b47cfe3e44fecfdd9e914ba99"
  dependencies:
    hyphenate-style-name "^1.0.2"
    isobject "^3.0.1"

fast-loops@^1.0.0:
  version "1.0.0"
  resolved "https://registry.yarnpkg.com/fast-loops/-/fast-loops-1.0.0.tgz#2cdd7e0ff67343b2b5f5e627d855a50b4bed559a"

fela-tools@^5.1.4:
  version "5.1.5"
  resolved "https://registry.yarnpkg.com/fela-tools/-/fela-tools-5.1.5.tgz#f8f845daab2b22da62cad965f868f945d2606c73"
  dependencies:
    css-in-js-utils "2.0.0"
    fast-loops "^1.0.0"
    fela "^6.1.7"
    fela-utils "^8.0.6"

fela-utils@^8.0.5, fela-utils@^8.0.6:
  version "8.0.6"
  resolved "https://registry.yarnpkg.com/fela-utils/-/fela-utils-8.0.6.tgz#f88bbf26f5d6bced52389fcac9a8fa4be82c0933"
  dependencies:
    css-in-js-utils "^2.0.0"
    fast-loops "^1.0.0"
    string-hash "^1.1.3"

fela@^6.1.7:
  version "6.1.7"
  resolved "https://registry.yarnpkg.com/fela/-/fela-6.1.7.tgz#be82e46d7337452a143e6504314c8015b9a9512d"
  dependencies:
    css-in-js-utils "2.0.0"
    fast-loops "^1.0.0"
    fela-utils "^8.0.6"
    isobject "^3.0.1"

hyphenate-style-name@^1.0.2:
  version "1.0.2"
  resolved "https://registry.yarnpkg.com/hyphenate-style-name/-/hyphenate-style-name-1.0.2.tgz#31160a36930adaf1fc04c6074f7eb41465d4ec4b"

isobject@^3.0.1:
  version "3.0.1"
  resolved "https://registry.yarnpkg.com/isobject/-/isobject-3.0.1.tgz#4e431e92b11a9731636aa1f9c8d1ccbcfdab78df"

string-hash@^1.1.3:
  version "1.1.3"
  resolved "https://registry.yarnpkg.com/string-hash/-/string-hash-1.1.3.tgz#e8aafc0ac1855b4666929ed7dd1275df5d6c811b"
