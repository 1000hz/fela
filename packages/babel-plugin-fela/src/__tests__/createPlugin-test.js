import { transformFileSync } from 'babel-core'
import { createRenderer } from 'fela'
import webPreset from 'fela-preset-web'

import createPlugin from '../createPlugin'

function transformFile(filename, plugin) {
  return transformFileSync(`${__dirname}/__fixtures__/${filename}.js`, {
    plugins: [plugin],
    babelrc: false
  }).code
}

describe('Using babel-plugin-fela', () => {
  it('should prerender static styles as a separate rule', () => {
    const plugin = createPlugin()

    expect(transformFile('createComponent', plugin)).toMatchSnapshot()
  })

  it('should precompile static styles', () => {
    const renderer = createRenderer({
      plugins: [...webPreset]
    })

    const plugin = createPlugin({
      precompile: true,
      renderer
    })

    expect(transformFile('createComponent', plugin)).toMatchSnapshot()
  })
})
