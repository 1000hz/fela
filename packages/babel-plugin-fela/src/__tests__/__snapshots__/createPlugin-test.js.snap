// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Using babel-plugin-fela should precompile static styles 1`] = `
"import { createComponent } from 'react-fela';

const rule = ({ fontSize, padding }, renderer) => {
  if (!renderer.cache.abcd) {
    renderer.cache.abcd = {
      type: 'PRECOMPILATION',
      className: 'a b c d'
    };
    if (!renderer.cache['colorred']) renderer.cache['colorred'] = {
      type: 'RULE',
      className: 'a',
      selector: '.a',
      declaration: 'color:red',
      media: '',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px)backgroundColorred']) renderer.cache['(min-height: 300px)backgroundColorred'] = {
      type: 'RULE',
      className: 'b',
      selector: '.b',
      declaration: 'background-color:red',
      media: '(min-height: 300px)',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px)lineHeight2']) renderer.cache['(min-height: 300px)lineHeight2'] = {
      type: 'RULE',
      className: 'c',
      selector: '.c',
      declaration: 'line-height:2',
      media: '(min-height: 300px)',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px):hovercolorblack']) renderer.cache['(min-height: 300px):hovercolorblack'] = {
      type: 'RULE',
      className: 'd',
      selector: '.d:hover',
      declaration: 'color:black',
      media: '(min-height: 300px)',
      support: ''
    };
  }

  return {
    _className: renderer.cache.abcd.className,

    fontSize,

    '@media (min-height: 300px)': {
      ':hover': {
        paddingLeft: padding
      }
    }
  };
};

export default createComponent(rule);"
`;

exports[`Using babel-plugin-fela should prerender static styles as a separate rule 1`] = `
"import { createComponent } from 'react-fela';

const rule = ({ fontSize, padding }, renderer) => {
  if (!renderer.cache.fktg8sm) {
    renderer.cache.fktg8sm = {
      type: 'PRERENDERING',
      className: renderer.renderRule(() => ({
        color: 'red',
        '@media (min-height: 300px)': {
          backgroundColor: 'red',
          lineHeight: 2,
          ':hover': {
            color: 'black'
          }
        }
      }))
    };
  }

  return {
    _className: renderer.cache.fktg8sm.className,

    fontSize, '@media (min-height: 300px)': { ':hover': { paddingLeft: padding
      }
    }
  };
};

export default createComponent(rule);"
`;
